File: backend/src/modules/ai-learning/ai-learning.service.ts
import { Injectable, NotFoundException, BadRequestException } from '@nestjs/common';
import { PrismaService } from '../../prisma/prisma.service';

@Injectable()
export class AILearningService {
  constructor(private prisma: PrismaService) {}

  async getLearningProfile(userId: string) {
    let profile = await this.prisma.learningProfile.findUnique({
      where: { userId },
    });

    if (!profile) {
      profile = await this.prisma.learningProfile.create({
        data: {
          userId,
          currentLevel: 'A1',
          learningGoal: 'conversation',
          dailyMinutes: 15,
          weakAreas: [],
          strongAreas: [],
          preferredTopics: ['Greetings', 'Numbers', 'Colors'],
          learningSpeed: 'normal',
          lastAnalysisDate: new Date(),
          totalStudyHours: 0,
          overallProgress: 0,
        },
      });
    }

    return profile;
  }

  async updateLearningProfile(userId: string, data: {
    currentLevel?: string;
    learningGoal?: string;
    dailyMinutes?: number;
    learningSpeed?: string;
    preferredTopics?: string[];
  }) {
    return this.prisma.learningProfile.upsert({
      where: { userId },
      update: data,
      create: {
        userId,
        currentLevel: data.currentLevel || 'A1',
        learningGoal: data.learningGoal || 'conversation',
        dailyMinutes: data.dailyMinutes || 15,
        learningSpeed: data.learningSpeed || 'normal',
        preferredTopics: data.preferredTopics || ['Greetings'],
        lastAnalysisDate: new Date(),
      },
    });
  }

  async analyzePerformance(userId: string, data: {
    quizResults: Array<{ category: string; correct: boolean; mistake?: string }>;
    flashcardResults: Array<{ word: string; category: string; remembered: boolean }>;
    speakingResults: Array<{ category: string; score: number; mistakes: string[] }>;
    aiConversationMistakes: Array<{ category: string; error: string }>;
  }) {
    const profile = await this.getLearningProfile(userId);
    const weakAreas = profile.weakAreas as any[] || [];
    const strongAreas = profile.strongAreas as any[] || [];

    const weaknessMap = new Map<string, any>();
    const strengthMap = new Map<string, any>();

    for (const quiz of data.quizResults) {
      if (quiz.correct) {
        const existing = strengthMap.get(quiz.category) || { correctCount: 0, mastery: 0 };
        strengthMap.set(quiz.category, {
          ...existing,
          category: quiz.category,
          correctCount: existing.correctCount + 1,
          mastery: Math.min(1, existing.mastery + 0.05),
          lastPracticedAt: new Date(),
        });
      } else if (quiz.mistake) {
        const existing = weaknessMap.get(quiz.category) || { mistakeCount: 0, severity: 0, commonMistakes: [] };
        weaknessMap.set(quiz.category, {
          ...existing,
          category: quiz.category,
          mistakeCount: existing.mistakeCount + 1,
          severity: Math.min(1, existing.severity + 0.05),
          lastMistakeAt: new Date(),
          commonMistakes: [...new Set([...existing.commonMistakes, quiz.mistake])].slice(0, 5),
        });
      }
    }

    for (const card of data.flashcardResults) {
      if (card.remembered) {
        const existing = strengthMap.get(card.category) || { correctCount: 0, mastery: 0 };
        strengthMap.set(card.category, {
          ...existing,
          category: card.category,
          correctCount: existing.correctCount + 1,
          mastery: Math.min(1, existing.mastery + 0.05),
          lastPracticedAt: new Date(),
        });
      } else {
        const existing = weaknessMap.get(card.category) || { mistakeCount: 0, severity: 0, commonMistakes: [] };
        weaknessMap.set(card.category, {
          ...existing,
          category: card.category,
          mistakeCount: existing.mistakeCount + 1,
          severity: Math.min(1, existing.severity + 0.05),
          lastMistakeAt: new Date(),
          commonMistakes: [...new Set([...existing.commonMistakes, `Forgot: ${card.word}`])].slice(0, 5),
        });
      }
    }

    for (const speaking of data.speakingResults) {
      if (speaking.score >= 0.8) {
        const existing = strengthMap.get(speaking.category) || { correctCount: 0, mastery: 0 };
        strengthMap.set(speaking.category, {
          ...existing,
          category: speaking.category,
          correctCount: existing.correctCount + 1,
          mastery: Math.min(1, existing.mastery + 0.05),
          lastPracticedAt: new Date(),
        });
      } else {
        for (const mistake of speaking.mistakes) {
          const existing = weaknessMap.get(speaking.category) || { mistakeCount: 0, severity: 0, commonMistakes: [] };
          weaknessMap.set(speaking.category, {
            ...existing,
            category: speaking.category,
            mistakeCount: existing.mistakeCount + 1,
            severity: Math.min(1, existing.severity + 0.05),
            lastMistakeAt: new Date(),
            commonMistakes: [...new Set([...existing.commonMistakes, mistake])].slice(0, 5),
          });
        }
      }
    }

    for (const mistake of data.aiConversationMistakes) {
      const existing = weaknessMap.get(mistake.category) || { mistakeCount: 0, severity: 0, commonMistakes: [] };
      weaknessMap.set(mistake.category, {
        ...existing,
        category: mistake.category,
        mistakeCount: existing.mistakeCount + 1,
        severity: Math.min(1, existing.severity + 0.05),
        lastMistakeAt: new Date(),
        commonMistakes: [...new Set([...existing.commonMistakes, mistake.error])].slice(0, 5),
      });
    }

    const sortedWeaknesses = Array.from(weaknessMap.values())
      .sort((a, b) => b.severity - a.severity)
      .slice(0, 5);

    const sortedStrengths = Array.from(strengthMap.values())
      .sort((a, b) => b.mastery - a.mastery)
      .slice(0, 5);

    const totalAttempts = data.quizResults.length + data.flashcardResults.length;
    const correctAttempts = data.quizResults.filter(q => q.correct).length +
        data.flashcardResults.filter(c => c.remembered).length;
    const progress = totalAttempts > 0 ? correctAttempts / totalAttempts : 0;

    return this.prisma.learningProfile.update({
      where: { userId },
      data: {
        weakAreas: sortedWeaknesses,
        strongAreas: sortedStrengths,
        lastAnalysisDate: new Date(),
        overallProgress: progress,
      },
    });
  }

  async getRecommendations(userId: string) {
    const profile = await this.getLearningProfile(userId);
    const weakAreas = profile.weakAreas as any[] || [];
    const strongAreas = profile.strongAreas as any[] || [];
    
    const recommendations = [];
    let priority = 10;

    for (const weakness of weakAreas.slice(0, 3)) {
      recommendations.push({
        id: `rec_weak_${weakness.category}_${Date.now()}`,
        type: 'weakness_fix',
        title: `Practice ${weakness.category}`,
        description: this.getWeaknessDescription(weakness),
        category: weakness.category,
        estimatedMinutes: this.estimateMinutes(weakness.severity),
        priority: priority--,
        isCompleted: false,
      });
    }

    if (strongAreas.length > 0) {
      const weakestStrong = strongAreas[strongAreas.length - 1];
      if (weakestStrong.mastery < 0.8) {
        recommendations.push({
          id: `rec_review_${Date.now()}`,
          type: 'review',
          title: `Review ${weakestStrong.category}`,
          description: `Maintain your progress in ${weakestStrong.category} with a quick review session.`,
          category: weakestStrong.category,
          estimatedMinutes: 10,
          priority: priority--,
          isCompleted: false,
        });
      }
    }

    recommendations.push({
      id: `rec_daily_${Date.now()}`,
      type: 'daily',
      title: this.getDailyTitle(profile),
      description: this.getDailyDescription(profile),
      category: 'Daily',
      estimatedMinutes: profile.dailyMinutes as number,
      priority: 5,
      isCompleted: false,
    });

    return recommendations;
  }

  async generateStudyPlan(userId: string) {
    const profile = await this.getLearningProfile(userId);
    const dayNames = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
    const activities = this.getActivitiesForGoal(profile.learningGoal as string);
    
    const days = dayNames.map((dayName, dayIndex) => {
      const dayActivities = [];
      let remainingMinutes = profile.dailyMinutes as number;
      let activityIndex = dayIndex * 2;

      while (remainingMinutes > 0 && dayActivities.length < 4) {
        const activity = activities[activityIndex % activities.length];
        const minutes = Math.min(activity.minutes, remainingMinutes);
        dayActivities.push({
          ...activity,
          minutes,
        });
        remainingMinutes -= minutes;
        activityIndex++;
      }

      return {
        dayName,
        activities: dayActivities,
        totalMinutes: (profile.dailyMinutes as number) - remainingMinutes,
      };
    });

    const plan = await this.prisma.studyPlan.create({
      data: {
        userId,
        title: `${this.getGoalTitle(profile.learningGoal as string)} - ${profile.currentLevel} Level`,
        startDate: new Date(),
        endDate: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000),
        days,
        isActive: true,
      },
    });

    return plan;
  }

  async getStudyPlan(userId: string) {
    return this.prisma.studyPlan.findFirst({
      where: { userId, isActive: true },
      orderBy: { createdAt: 'desc' },
    });
  }

  async getAnalytics() {
    const [
      totalProfiles,
      levelDistribution,
      goalDistribution,
      avgProgress,
      avgDailyMinutes,
    ] = await Promise.all([
      this.prisma.learningProfile.count(),
      this.prisma.learningProfile.groupBy({
        by: ['currentLevel'],
        _count: true,
      }),
      this.prisma.learningProfile.groupBy({
        by: ['learningGoal'],
        _count: true,
      }),
      this.prisma.learningProfile.aggregate({
        _avg: { overallProgress: true },
      }),
      this.prisma.learningProfile.aggregate({
        _avg: { dailyMinutes: true },
      }),
    ]);

    const weakTopics = await this.prisma.$queryRaw`
      SELECT 
        jsonb_array_elements("weakAreas")->>'category' as category,
        COUNT(*) as count
      FROM "LearningProfile"
      WHERE "weakAreas" != '[]'::jsonb
      GROUP BY category
      ORDER BY count DESC
      LIMIT 10
    `;

    return {
      totalProfiles,
      levelDistribution,
      goalDistribution,
      avgProgress: avgProgress._avg.overallProgress || 0,
      avgDailyMinutes: avgDailyMinutes._avg.dailyMinutes || 15,
      weakTopics,
    };
  }

  private getWeaknessDescription(weakness: any): string {
    const descriptions: Record<string, string> = {
      'Articles': `You've made ${weakness.mistakeCount} mistakes with German articles. Focus on der/die/das patterns.`,
      'Verb Conjugation': 'Practice verb conjugations to improve your sentence structure.',
      'Cases': 'Work on Nominative, Accusative, Dative, and Genitive cases.',
      'Word Order': 'German word order can be tricky. Practice placing verbs correctly.',
      'Vocabulary': 'Review vocabulary in this category to strengthen your word bank.',
      'Pronunciation': 'Practice pronunciation to sound more natural in German.',
      'Listening': 'Improve your listening skills with targeted exercises.',
    };
    return descriptions[weakness.category] || `Focus on ${weakness.category} to improve your overall skills.`;
  }

  private estimateMinutes(severity: number): number {
    if (severity >= 0.8) return 20;
    if (severity >= 0.6) return 15;
    if (severity >= 0.4) return 10;
    return 5;
  }

  private getDailyTitle(goal: string): string {
    const titles: Record<string, string> = {
      'travel': 'Practice Travel Phrases!',
      'work': 'Business German Time!',
      'study': 'Academic Practice Session',
      'conversation': 'Daily Conversation Practice',
      'hobby': 'Explore German Culture!',
      'goetheExam': 'Exam Prep Time!',
    };
    return titles[goal] || 'Keep Learning German!';
  }

  private getDailyDescription(profile: any): string {
    const weakAreas = profile.weakAreas as any[] || [];
    if (weakAreas.length > 0) {
      return `Spend ${profile.dailyMinutes} minutes practicing ${weakAreas[0].category}. You're making progress!`;
    }
    return `Practice German for ${profile.dailyMinutes} minutes today to maintain your streak!`;
  }

  private getActivitiesForGoal(goal: string): any[] {
    const activities: Record<string, any[]> = {
      'goetheExam': [
        { type: 'grammar', title: 'Grammar Focus', minutes: 10, description: 'Practice exam-style grammar' },
        { type: 'vocabulary', title: 'Vocabulary Builder', minutes: 10, description: 'Learn exam-level vocabulary' },
        { type: 'listening', title: 'Listening Comprehension', minutes: 10, description: 'Practice with exam audio' },
        { type: 'quiz', title: 'Practice Quiz', minutes: 10, description: 'Test your knowledge' },
      ],
      'work': [
        { type: 'vocabulary', title: 'Business Vocabulary', minutes: 10, description: 'Learn professional terms' },
        { type: 'ai_chat', title: 'Meeting Practice', minutes: 10, description: 'Practice business conversations' },
        { type: 'grammar', title: 'Formal Grammar', minutes: 10, description: 'Master formal expressions' },
      ],
      'travel': [
        { type: 'vocabulary', title: 'Travel Phrases', minutes: 10, description: 'Essential travel vocabulary' },
        { type: 'ai_chat', title: 'Hotel & Restaurant', minutes: 10, description: 'Practice booking and ordering' },
        { type: 'speaking', title: 'Pronunciation', minutes: 10, description: 'Sound natural when traveling' },
      ],
      'default': [
        { type: 'vocabulary', title: 'Daily Vocabulary', minutes: 10, description: 'Build your word bank' },
        { type: 'grammar', title: 'Grammar Practice', minutes: 10, description: 'Strengthen grammar foundation' },
        { type: 'ai_chat', title: 'Conversation Practice', minutes: 10, description: 'Chat with AI in German' },
        { type: 'listening', title: 'Listening Exercise', minutes: 10, description: 'Improve comprehension' },
      ],
    };
    return activities[goal] || activities['default'];
  }

  private getGoalTitle(goal: string): string {
    const titles: Record<string, string> = {
      'travel': 'Travel & Tourism',
      'work': 'Business German',
      'study': 'Academic Studies',
      'conversation': 'Daily Conversation',
      'hobby': 'Hobby & Culture',
      'goetheExam': 'Goethe Exam Preparation',
    };
    return titles[goal] || 'General Learning';
  }
}

File: lib/features/ai_learning/ai_learning_service.dart
import 'dart:math';
import 'package:lexi/features/ai_learning/models/learning_profile.dart';

class AILearningService {
  final Random _random = Random();
  LearningProfile generateDefaultProfile(String userId) {
    return LearningProfile(
      userId: userId,
      currentLevel: 'A1',
      learningGoal: LearningGoal.conversation,
      dailyMinutes: 15,
      weakAreas: [],
      strongAreas: [],
      preferredTopics: ['Greetings', 'Numbers', 'Colors'],
      learningSpeed: LearningSpeed.normal,
      lastAnalysisDate: DateTime.now(),
      totalStudyHours: 0,
      currentStreak: 0,
      overallProgress: 0.0,
    );
  }

  LearningProfile analyzeAndUpdateProfile(
    LearningProfile profile, {
    required List<Map<String, dynamic>> quizResults,
    required List<Map<String, dynamic>> flashcardResults,
    required List<Map<String, dynamic>> speakingResults,
    required List<Map<String, dynamic>> aiConversationMistakes,
  }) {
    final Map<String, WeaknessArea> weaknesses = {};
    final Map<String, StrongArea> strengths = {};
    for (final quiz in quizResults) {
      final category = quiz['category'] ?? 'General';
      final correct = quiz['correct'] as bool? ?? false;
      final mistake = quiz['mistake'] as String?;
      if (correct) {
        strengths[category] = _updateStrongArea(
          strengths[category],
          category,
          'quiz',
        );
      } else if (mistake != null) {
        weaknesses[category] = _updateWeaknessArea(
          weaknesses[category],
          category,
          mistake,
        );
      }
    }
    for (final card in flashcardResults) {
      final word = card['word'] ?? '';
      final category = card['category'] ?? 'Vocabulary';
      final remembered = card['remembered'] as bool? ?? false;
      if (remembered) {
        strengths[category] = _updateStrongArea(
          strengths[category],
          category,
          'flashcard',
        );
      } else {
        weaknesses[category] = _updateWeaknessArea(
          weaknesses[category],
          category,
          'Forgot: $word',
        );
      }
    }
    for (final speaking in speakingResults) {
      final category = speaking['category'] ?? 'Speaking';
      final score = speaking['score'] as double? ?? 0.0;
      final mistakes = List<String>.from(speaking['mistakes'] ?? []);
      if (score >= 0.8) {
        strengths[category] = _updateStrongArea(
          strengths[category],
          category,
          'speaking',
        );
      } else {
        for (final mistake in mistakes) {
          weaknesses[category] = _updateWeaknessArea(
            weaknesses[category],
            category,
            mistake,
          );
        }
      }
    }
    for (final mistake in aiConversationMistakes) {
      final category = mistake['category'] ?? 'Grammar';
      final error = mistake['error'] ?? '';
      weaknesses[category] = _updateWeaknessArea(
        weaknesses[category],
        category,
        error,
      );
    }
    final sortedWeaknesses = weaknesses.values.toList()
      ..sort((a, b) => b.severity.compareTo(a.severity));
    final sortedStrengths = strengths.values.toList()
      ..sort((a, b) => b.mastery.compareTo(a.mastery));
    final totalAttempts = quizResults.length + flashcardResults.length;
    final correctAttempts =
        quizResults.where((q) => q['correct'] == true).length +
        flashcardResults.where((c) => c['remembered'] == true).length;
    final progress = totalAttempts > 0 ? correctAttempts / totalAttempts : 0.0;
    return profile.copyWith(
      weakAreas: sortedWeaknesses.take(5).toList(),
      strongAreas: sortedStrengths.take(5).toList(),
      lastAnalysisDate: DateTime.now(),
      overallProgress: progress,
    );
  }

  WeaknessArea _updateWeaknessArea(
    WeaknessArea? existing,
    String category,
    String mistake,
  ) {
    if (existing != null) {
      final newMistakeCount = existing.mistakeCount + 1;
      final newSeverity = min(1.0, existing.severity + 0.05);
      final newMistakes = List<String>.from(existing.commonMistakes);
      if (!newMistakes.contains(mistake) && newMistakes.length < 5) {
        newMistakes.add(mistake);
      }
      return WeaknessArea(
        category: category,
        subCategory: existing.subCategory,
        mistakeCount: newMistakeCount,
        severity: newSeverity,
        lastMistakeAt: DateTime.now(),
        commonMistakes: newMistakes,
      );
    }
    return WeaknessArea(
      category: category,
      subCategory: 'General',
      mistakeCount: 1,
      severity: 0.3,
      lastMistakeAt: DateTime.now(),
      commonMistakes: [mistake],
    );
  }

  StrongArea _updateStrongArea(
    StrongArea? existing,
    String category,
    String source,
  ) {
    if (existing != null) {
      final newCorrectCount = existing.correctCount + 1;
      final newMastery = min(1.0, existing.mastery + 0.05);
      return StrongArea(
        category: category,
        subCategory: existing.subCategory,
        correctCount: newCorrectCount,
        mastery: newMastery,
        lastPracticedAt: DateTime.now(),
      );
    }
    return StrongArea(
      category: category,
      subCategory: 'General',
      correctCount: 1,
      mastery: 0.3,
      lastPracticedAt: DateTime.now(),
    );
  }

  List<AIRecommendation> generateRecommendations(LearningProfile profile) {
    final recommendations = <AIRecommendation>[];
    int priority = 10;
    for (final weakness in profile.weakAreas.take(3)) {
      recommendations.add(
        AIRecommendation(
          id: 'rec_weak_${weakness.category}_${DateTime.now().millisecondsSinceEpoch}',
          userId: profile.userId,
          type: 'weakness_fix',
          title: 'Practice ${weakness.category}',
          description: _getWeaknessDescription(weakness),
          category: weakness.category,
          estimatedMinutes: _estimateMinutes(weakness.severity),
          priority: priority--,
          createdAt: DateTime.now(),
        ),
      );
    }
    if (profile.strongAreas.isNotEmpty) {
      final weakestStrong = profile.strongAreas.last;
      if (weakestStrong.mastery < 0.8) {
        recommendations.add(
          AIRecommendation(
            id: 'rec_review_${DateTime.now().millisecondsSinceEpoch}',
            userId: profile.userId,
            type: 'review',
            title: 'Review ${weakestStrong.category}',
            description:
                'Maintain your progress in ${weakestStrong.category} with a quick review session.',
            category: weakestStrong.category,
            estimatedMinutes: 10,
            priority: priority--,
            createdAt: DateTime.now(),
          ),
        );
      }
    }
    final newTopics = _getNewTopicsForLevel(
      profile.currentLevel,
      profile.weakAreas,
    );
    if (newTopics.isNotEmpty) {
      recommendations.add(
        AIRecommendation(
          id: 'rec_new_${DateTime.now().millisecondsSinceEpoch}',
          userId: profile.userId,
          type: 'new_topic',
          title: 'Explore: ${newTopics.first}',
          description:
              'Ready to learn something new? Try ${newTopics.first} to expand your knowledge.',
          category: 'New Content',
          estimatedMinutes: 15,
          priority: priority--,
          createdAt: DateTime.now(),
        ),
      );
    }
    if (profile.currentStreak > 0 && profile.currentStreak % 7 == 0) {
      recommendations.add(
        AIRecommendation(
          id: 'rec_challenge_${DateTime.now().millisecondsSinceEpoch}',
          userId: profile.userId,
          type: 'challenge',
          title: 'Weekly Challenge',
          description:
              'You\'ve been consistent for ${profile.currentStreak} days! Take on a challenge to test your skills.',
          category: 'Challenge',
          estimatedMinutes: 20,
          priority: 8,
          createdAt: DateTime.now(),
        ),
      );
    }
    recommendations.add(
      AIRecommendation(
        id: 'rec_daily_${DateTime.now().millisecondsSinceEpoch}',
        userId: profile.userId,
        type: 'daily',
        title: _getDailyRecommendationTitle(profile),
        description: _getDailyRecommendation(profile),
        category: 'Daily',
        estimatedMinutes: profile.dailyMinutes,
        priority: 5,
        createdAt: DateTime.now(),
      ),
    );
    return recommendations;
  }

  String _getWeaknessDescription(WeaknessArea weakness) {
    switch (weakness.category.toLowerCase()) {
      case 'articles':
        return 'You\'ve made ${weakness.mistakeCount} mistakes with German articles. Focus on der/die/das patterns.';
      case 'verb conjugation':
        return 'Practice verb conjugations to improve your sentence structure.';
      case 'cases':
        return 'Work on Nominative, Accusative, Dative, and Genitive cases.';
      case 'word order':
        return 'German word order can be tricky. Practice placing verbs correctly.';
      case 'vocabulary':
        return 'Review vocabulary in this category to strengthen your word bank.';
      case 'pronunciation':
        return 'Practice pronunciation to sound more natural in German.';
      case 'listening':
        return 'Improve your listening skills with targeted exercises.';
      default:
        return 'Focus on ${weakness.category} to improve your overall skills.';
    }
  }

  int _estimateMinutes(double severity) {
    if (severity >= 0.8) return 20;
    if (severity >= 0.6) return 15;
    if (severity >= 0.4) return 10;
    return 5;
  }

  List<String> _getNewTopicsForLevel(
    String level,
    List<WeaknessArea> weaknesses,
  ) {
    final allTopics = {
      'A1': ['Greetings', 'Numbers', 'Colors', 'Family', 'Food', 'Body Parts'],
      'A2': [
        'Shopping',
        'Travel',
        'Health',
        'Weather',
        'Hobbies',
        'Directions',
      ],
      'B1': [
        'Work',
        'Education',
        'Environment',
        'Technology',
        'Culture',
        'News',
      ],
      'B2': ['Politics', 'Science', 'Arts', 'Philosophy', 'Economics', 'Law'],
      'C1': [
        'Academic Writing',
        'Professional',
        'Literature',
        'Media',
        'Research',
      ],
      'C2': ['Idioms', 'Dialects', 'Advanced Grammar', 'Nuanced Expression'],
    };
    final currentTopics = allTopics[level] ?? allTopics['A1']!;
    final weakCategories = weaknesses
        .map((w) => w.category.toLowerCase())
        .toList();
    return currentTopics
        .where((topic) => !weakCategories.contains(topic.toLowerCase()))
        .toList();
  }

  String _getDailyRecommendationTitle(LearningProfile profile) {
    final titles = [
      'Start your German journey today!',
      'Time for a quick German practice!',
      'Keep your streak alive!',
      'Challenge yourself with German!',
      'Learn something new in German!',
    ];
    return titles[_random.nextInt(titles.length)];
  }

  String _getDailyRecommendation(LearningProfile profile) {
    if (profile.weakAreas.isNotEmpty) {
      final weakest = profile.weakAreas.first;
      return 'Spend ${profile.dailyMinutes} minutes practicing ${weakest.category}. You\'re making progress!';
    }
    return 'Practice German for ${profile.dailyMinutes} minutes today to maintain your streak!';
  }

  StudyPlan generateStudyPlan(LearningProfile profile) {
    final days = <StudyPlanDay>[];
    final dayNames = [
      'Monday',
      'Tuesday',
      'Wednesday',
      'Thursday',
      'Friday',
      'Saturday',
      'Sunday',
    ];
    final activities = _getActivitiesForGoal(profile);
    int activityIndex = 0;
    for (final dayName in dayNames) {
      final dayActivities = <StudyPlanActivity>[];
      int remainingMinutes = profile.dailyMinutes;
      while (remainingMinutes > 0 && activityIndex < activities.length * 2) {
        final activity = activities[activityIndex % activities.length];
        final minutes = min(activity.minutes, remainingMinutes);
        dayActivities.add(
          StudyPlanActivity(
            type: activity.type,
            title: activity.title,
            minutes: minutes,
            description: activity.description,
          ),
        );
        remainingMinutes -= minutes;
        activityIndex++;
      }
      days.add(
        StudyPlanDay(
          dayName: dayName,
          activities: dayActivities,
          totalMinutes: profile.dailyMinutes - remainingMinutes,
        ),
      );
    }
    final goalTitle = profile.goalText;
    return StudyPlan(
      id: 'plan_${DateTime.now().millisecondsSinceEpoch}',
      userId: profile.userId,
      title: '$goalTitle - ${profile.currentLevel} Level',
      startDate: DateTime.now(),
      endDate: DateTime.now().add(const Duration(days: 7)),
      days: days,
      isActive: true,
      createdAt: DateTime.now(),
    );
  }

  List<StudyPlanActivity> _getActivitiesForGoal(LearningProfile profile) {
    switch (profile.learningGoal) {
      case LearningGoal.goetheExam:
        return [
          const StudyPlanActivity(
            type: 'grammar',
            title: 'Grammar Focus',
            minutes: 10,
            description: 'Practice exam-style grammar questions',
          ),
          const StudyPlanActivity(
            type: 'vocabulary',
            title: 'Vocabulary Builder',
            minutes: 10,
            description: 'Learn exam-level vocabulary',
          ),
          const StudyPlanActivity(
            type: 'listening',
            title: 'Listening Comprehension',
            minutes: 10,
            description: 'Practice with exam audio',
          ),
          const StudyPlanActivity(
            type: 'quiz',
            title: 'Practice Quiz',
            minutes: 10,
            description: 'Test your knowledge',
          ),
          const StudyPlanActivity(
            type: 'ai_chat',
            title: 'Writing Practice',
            minutes: 10,
            description: 'Practice essay writing with AI',
          ),
        ];
      case LearningGoal.work:
        return [
          const StudyPlanActivity(
            type: 'vocabulary',
            title: 'Business Vocabulary',
            minutes: 10,
            description: 'Learn professional terms',
          ),
          const StudyPlanActivity(
            type: 'ai_chat',
            title: 'Meeting Practice',
            minutes: 10,
            description: 'Practice business conversations',
          ),
          const StudyPlanActivity(
            type: 'grammar',
            title: 'Formal Grammar',
            minutes: 10,
            description: 'Master formal expressions',
          ),
          const StudyPlanActivity(
            type: 'listening',
            title: 'Business Audio',
            minutes: 10,
            description: 'Listen to business German',
          ),
        ];
      case LearningGoal.travel:
        return [
          const StudyPlanActivity(
            type: 'vocabulary',
            title: 'Travel Phrases',
            minutes: 10,
            description: 'Essential travel vocabulary',
          ),
          const StudyPlanActivity(
            type: 'ai_chat',
            title: 'Hotel & Restaurant',
            minutes: 10,
            description: 'Practice booking and ordering',
          ),
          const StudyPlanActivity(
            type: 'speaking',
            title: 'Pronunciation',
            minutes: 10,
            description: 'Sound natural when traveling',
          ),
          const StudyPlanActivity(
            type: 'listening',
            title: 'Train Announcements',
            minutes: 10,
            description: 'Understand public transport',
          ),
        ];
      default:
        return [
          const StudyPlanActivity(
            type: 'vocabulary',
            title: 'Daily Vocabulary',
            minutes: 10,
            description: 'Build your word bank',
          ),
          const StudyPlanActivity(
            type: 'grammar',
            title: 'Grammar Practice',
            minutes: 10,
            description: 'Strengthen your grammar foundation',
          ),
          const StudyPlanActivity(
            type: 'ai_chat',
            title: 'Conversation Practice',
            minutes: 10,
            description: 'Chat with AI in German',
          ),
          const StudyPlanActivity(
            type: 'listening',
            title: 'Listening Exercise',
            minutes: 10,
            description: 'Improve your comprehension',
          ),
          const StudyPlanActivity(
            type: 'quiz',
            title: 'Daily Quiz',
            minutes: 10,
            description: 'Test what you learned',
          ),
        ];
    }
  }

  StudentMemory updateStudentMemory(
    StudentMemory memory, {
    String? newMistake,
    String? mistakeCategory,
    String? conversationTopic,
    List<String>? conversationMistakes,
    String? learnedWord,
  }) {
    final updatedPatterns = List<MistakePattern>.from(memory.mistakePatterns);
    final updatedLearned = List<String>.from(memory.successfullyLearned);
    final updatedConversations = List<ConversationSummary>.from(
      memory.conversationHistory,
    );
    if (newMistake != null && mistakeCategory != null) {
      final existingIndex = updatedPatterns.indexWhere(
        (p) => p.pattern == newMistake,
      );
      if (existingIndex != -1) {
        final existing = updatedPatterns[existingIndex];
        updatedPatterns[existingIndex] = MistakePattern(
          pattern: existing.pattern,
          category: existing.category,
          occurrences: existing.occurrences + 1,
          examples: existing.examples,
          firstSeen: existing.firstSeen,
          lastSeen: DateTime.now(),
        );
      } else {
        updatedPatterns.add(
          MistakePattern(
            pattern: newMistake,
            category: mistakeCategory,
            occurrences: 1,
            examples: [newMistake],
            firstSeen: DateTime.now(),
            lastSeen: DateTime.now(),
          ),
        );
      }
    }
    if (learnedWord != null && !updatedLearned.contains(learnedWord)) {
      updatedLearned.add(learnedWord);
    }
    if (conversationTopic != null) {
      updatedConversations.add(
        ConversationSummary(
          date: DateTime.now(),
          topic: conversationTopic,
          keyPoints: [],
          mistakesMade: conversationMistakes ?? [],
          durationMinutes: 5,
        ),
      );
      if (updatedConversations.length > 50) {
        updatedConversations.removeAt(0);
      }
    }
    return StudentMemory(
      userId: memory.userId,
      mistakePatterns: updatedPatterns,
      successfullyLearned: updatedLearned,
      conversationHistory: updatedConversations,
      preferences: memory.preferences,
      lastUpdated: DateTime.now(),
    );
  }
}

extension LearningProfileCopyWith on LearningProfile {
  LearningProfile copyWith({
    String? currentLevel,
    LearningGoal? learningGoal,
    int? dailyMinutes,
    List<WeaknessArea>? weakAreas,
    List<StrongArea>? strongAreas,
    List<String>? preferredTopics,
    LearningSpeed? learningSpeed,
    DateTime? lastAnalysisDate,
    int? totalStudyHours,
    int? currentStreak,
    double? overallProgress,
  }) {
    return LearningProfile(
      userId: userId,
      currentLevel: currentLevel ?? this.currentLevel,
      learningGoal: learningGoal ?? this.learningGoal,
      dailyMinutes: dailyMinutes ?? this.dailyMinutes,
      weakAreas: weakAreas ?? this.weakAreas,
      strongAreas: strongAreas ?? this.strongAreas,
      preferredTopics: preferredTopics ?? this.preferredTopics,
      learningSpeed: learningSpeed ?? this.learningSpeed,
      lastAnalysisDate: lastAnalysisDate ?? this.lastAnalysisDate,
      totalStudyHours: totalStudyHours ?? this.totalStudyHours,
      currentStreak: currentStreak ?? this.currentStreak,
      overallProgress: overallProgress ?? this.overallProgress,
    );
  }
}
